tests/test_utils/test_flash_attention.py

```diff
@@ -1,130 +1,257 @@
 
 
 @pytest.mark.skipif(not HAS_MEM_EFF_ATTN and not HAS_FLASH_ATTN, reason="xformers is not available")
 @clear_cache_before_run()
 @parameterize('proj_shape', [(6, 24, 8, 4, 16)])
 @parameterize('dtype', DTYPE)
-def test_cross_attention(proj_shape, dtype):
+@parameterize('dropout', [0.0, 0.1])
+def test_cross_attention(proj_shape, dtype, dropout):
     (B, S, T, H, D_HEAD) = proj_shape
     D = H * D_HEAD
 
-    q_attn = torch.nn.Linear(D, D, dtype=dtype, device="cuda")
-    kv_attn = torch.nn.Linear(D, 2 * D, dtype=dtype, device="cuda")
-
-    attn = ColoAttention(D, H, dropout=0.1)
+    q = torch.randn((B, T, H, D_HEAD), dtype=dtype, device="cuda", requires_grad=True)
+    k = torch.randn((B, S, H, D_HEAD), dtype=dtype, device="cuda", requires_grad=True)
+    v = torch.randn((B, S, H, D_HEAD), dtype=dtype, device="cuda", requires_grad=True)
 
-    src = torch.randn((B, S, D), dtype=dtype, device="cuda")
-    tgt = torch.randn((B, T, D), dtype=dtype, device="cuda")
-
-    q = q_attn(tgt)
-    kv = kv_attn(src)
-    q = rearrange(q, 'b s (h d) -> b s h d', h=H)
-    k, v = rearrange(kv, 'b s (n h d) -> b s n h d', n=2, h=H).unbind(dim=2)
+    attn = ColoAttention(D, H, dropout=dropout)
     y = attn(q, k, v, attn_mask_type=AttnMaskType.causal)
 
     assert list(y.shape) == [B, T, D]
 
-    dy = torch.rand_like(y)
-    y.backward(dy)
+    if dropout == 0.0:
+        out_ref = attention_ref(q, k, v, None, causal=True)
+        out_pt = attention_ref(q, k, v, None, causal=True, upcast=False, reorder_ops=True)
+
+        dy = torch.rand_like(y)
+        grad_q, grad_k, grad_v = torch.autograd.grad(y, (q, k, v), dy)
+        grad_ref_q, grad_ref_k, grad_ref_v = torch.autograd.grad(out_ref, (q, k, v), dy)
+        grad_pt_q, grad_pt_k, grad_pt_v = torch.autograd.grad(out_pt, (q, k, v), dy)
+
+        if dtype in FLASH_DTYPE:
+            assert (y - out_ref).abs().max().item() <= 2 * (out_pt - out_ref).abs().max().item()
+            assert (grad_q - grad_ref_q).abs().max().item() <= 2 * (grad_pt_q - grad_ref_q).abs().max().item()
+            assert (grad_k - grad_ref_k).abs().max().item() <= 2 * (grad_pt_k - grad_ref_k).abs().max().item()
+            assert (grad_v - grad_ref_v).abs().max().item() <= 2 * (grad_pt_v - grad_ref_v).abs().max().item()
+        else:
+            torch.allclose(y, out_ref, atol=1e-7), f"{(y - out_ref).abs().max()}"
+            torch.allclose(grad_q, grad_ref_q, atol=1e-7), f"{(grad_q - grad_ref_q).abs().max()}"
+            torch.allclose(grad_k, grad_ref_k, atol=1e-7), f"{(grad_k - grad_ref_k).abs().max()}"
+            torch.allclose(grad_v, grad_ref_v, atol=1e-7), f"{(grad_v - grad_ref_v).abs().max()}"
+
+
+@pytest.mark.skipif(not HAS_MEM_EFF_ATTN and not HAS_FLASH_ATTN, reason="xformers is not available")
+@clear_cache_before_run()
+@parameterize('proj_shape', [(6, 8, 4, 16)])
+@parameterize('dtype', DTYPE)
+def test_padding_process(proj_shape, dtype):
```

Comment:
integrate this check into previous checks, like make pad as a parameter and adjust attn mask & attn mask type accordingly